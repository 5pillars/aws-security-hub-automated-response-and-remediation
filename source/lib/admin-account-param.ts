// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export default class AdminAccountParam extends Construct {
  public readonly value: string;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const validAwsAccount = '^\\d{12}$';

    const param = new cdk.CfnParameter(this, 'Admin Account Number', {
      description: 'Admin account number',
      type: 'String',
      allowedPattern: validAwsAccount,
    });
    param.overrideLogicalId(`SecHubAdminAccount`);
    this.value = param.valueAsString;
  }
}
