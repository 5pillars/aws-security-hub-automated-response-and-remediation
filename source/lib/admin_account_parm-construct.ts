// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class AdminAccountParm extends Construct {
  public readonly adminAccountNumber: cdk.CfnParameter;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const validAwsAccount = '\\d{12}';

    this.adminAccountNumber = new cdk.CfnParameter(this, 'Admin Account Number', {
      description: 'Admin account number',
      type: 'String',
      allowedPattern: validAwsAccount,
    });
    this.adminAccountNumber.overrideLogicalId(`SecHubAdminAccount`);
  }
}
