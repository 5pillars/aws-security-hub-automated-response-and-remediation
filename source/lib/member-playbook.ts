// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CfnCondition, CfnParameter, CfnResource, Fn, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SerializedNestedStackFactory } from './cdk-helper/nested-stack';

export interface MemberPlaybookProps {
  readonly name: string;
  readonly nestedStackFactory: SerializedNestedStackFactory;
  readonly parameters?: { [_: string]: string };
  readonly stackDependencies?: CfnResource[];
  readonly defaultState?: 'yes' | 'no';
  readonly description?: string;
}

export class MemberPlaybookNew {
  parameterName = '';
  playbookStacks: Stack[];

  constructor(scope: Construct, props: MemberPlaybookProps) {
    const templateFileOriginal = `${props.name}MemberStackOriginal.template`;
    const templateFileNew = `${props.name}MemberStackNew.template`;
    const illegalChars = /[\\._]/g;
    const playbookName = props.name.replace(illegalChars, '');
    this.parameterName = `Load${playbookName}MemberStacks`;

    //---------------------------------------------------------------------
    // Playbook Template Nested Stack
    //
    const stackOptionOriginal = new CfnParameter(scope, `LoadOriginalMemberStack${playbookName}`, {
      type: 'String',
      description:
        props.description ?? `Install the member components for automated remediation of ${props.name} controls?`,
      default: props.defaultState ?? 'no',
      allowedValues: ['yes', 'no'],
    });
    const stackOptionNew = new CfnParameter(scope, `LoadNewMemberStack${playbookName}`, {
      type: 'String',
      description:
        props.description ?? `Install the member components for automated remediation of ${props.name} controls?`,
      default: props.defaultState ?? 'no',
      allowedValues: ['yes', 'no'],
    });
    stackOptionOriginal.overrideLogicalId(this.parameterName);
      this.playbookStacks = []
      const originalNestedStack = props.nestedStackFactory.addNestedStack(`PlaybookMemberStackOriginal${playbookName}`, {
        templateRelativePath: `playbooks/${templateFileOriginal}`,
        parameters: props.parameters,
        condition: new CfnCondition(scope, `load${playbookName}CondOriginal`, {
          expression: Fn.conditionEquals(stackOptionOriginal, 'yes'),
        }),
      });

    const cfnStackOriginal = originalNestedStack.nestedStackResource as CfnResource;
    cfnStackOriginal.overrideLogicalId(`PlaybookMemberStackOriginal${props.name}`);

    const newNestedStack = props.nestedStackFactory.addNestedStack(`PlaybookMemberStackNew${playbookName}`, {
      templateRelativePath: `playbooks/${templateFileNew}`,
      parameters: props.parameters,
      condition: new CfnCondition(scope, `load${playbookName}CondNew`, {
        expression: Fn.conditionEquals(stackOptionNew, 'yes'),
      }),
    });

    const cfnStackNew = newNestedStack.nestedStackResource as CfnResource;
    cfnStackNew.overrideLogicalId(`PlaybookMemberStackNew${props.name}`);

    this.playbookStacks.push(originalNestedStack as Stack)
    this.playbookStacks.push(newNestedStack as Stack)
  }
}

export class MemberPlaybook {
  parameterName = '';
  playbookStack: Stack;

  constructor(scope: Construct, props: MemberPlaybookProps) {
    const templateFile = `${props.name}MemberStack.template`;
    const illegalChars = /[\\._]/g;
    const playbookName = props.name.replace(illegalChars, '');
    this.parameterName = `Load${playbookName}MemberStack`;

    //---------------------------------------------------------------------
    // Playbook Template Nested Stack
    //
    const stackOption = new CfnParameter(scope, `LoadMemberStack${playbookName}`, {
      type: 'String',
      description:
        props.description ?? `Install the member components for automated remediation of ${props.name} controls?`,
      default: props.defaultState ?? 'no',
      allowedValues: ['yes', 'no'],
    });
    stackOption.overrideLogicalId(this.parameterName);
    this.playbookStack = props.nestedStackFactory.addNestedStack(`PlaybookMemberStack${playbookName}`, {
      templateRelativePath: `playbooks/${templateFile}`,
      parameters: props.parameters,
      condition: new CfnCondition(scope, `load${playbookName}Cond`, {
        expression: Fn.conditionEquals(stackOption, 'yes'),
      }),
    });
    const cfnStack = this.playbookStack.nestedStackResource as CfnResource;
    cfnStack.overrideLogicalId(`PlaybookMemberStack${props.name}`);
  }
}
