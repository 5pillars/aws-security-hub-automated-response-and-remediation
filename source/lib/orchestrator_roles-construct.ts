#!/usr/bin/env node
/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import { 
    PolicyStatement, 
    Effect, 
    Role, 
    PolicyDocument, 
    ArnPrincipal,
    ServicePrincipal,
    CompositePrincipal, 
    CfnRole 
} from '@aws-cdk/aws-iam';

export interface OrchRoleProps {
    solutionId: string;
    adminAccountId: string;
    adminRoleName: string;
}

export class OrchestratorMemberRole extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: OrchRoleProps) {
    super(scope, id);
    const RESOURCE_PREFIX = props.solutionId.replace(/^DEV-/,''); // prefix on every resource name
    const stack = cdk.Stack.of(this);
    const memberPolicy = new PolicyDocument();

    /**
     * @description Cross-account permissions for Orchestration role
     * @type {PolicyStatement}
     */
    const iamPerms = new PolicyStatement();
    iamPerms.addActions(
        "iam:PassRole",
        "iam:GetRole"
    )
    iamPerms.effect = Effect.ALLOW
    iamPerms.addResources(
        `arn:${stack.partition}:iam::${stack.account}:role/${RESOURCE_PREFIX}-*`
    );
    memberPolicy.addStatements(iamPerms)
    const ssmROPerms = new PolicyStatement()
    ssmROPerms.addActions(
        "ssm:DescribeAutomationExecutions",
        "ssm:DescribeDocument",
        "ssm:GetParameters"
    )
    ssmROPerms.effect = Effect.ALLOW;
    ssmROPerms.addResources(
        "arn:" + stack.partition + ":ssm:*:*:*"
    )
    memberPolicy.addStatements(ssmROPerms)

    const ssmRWPerms = new PolicyStatement()
    ssmRWPerms.addActions(
        "ssm:StartAutomationExecution",
        "ssm:GetAutomationExecution"
    )
    ssmRWPerms.addResources(
        // `arn:${stack.partition}:ssm:*:${stack.account}:document/SHARR-*`,
        // `arn:${stack.partition}:ssm:*:${stack.account}:automation-definition/*`,
        // `arn:${stack.partition}:ssm:*:${stack.account}:document/SHARR-*`,
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'document',
            resourceName: 'SHARR-*',
            sep: '/'
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-definition',
            resourceName: '*',
            sep: '/'
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-definition',
            account:'',
            resourceName: '*',
            sep: '/'
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-execution',
            resourceName: '*',
            sep: '/'
        })
    );
    memberPolicy.addStatements(ssmRWPerms)

    const sechubPerms = new PolicyStatement();
    sechubPerms.addActions("cloudwatch:PutMetricData")
    sechubPerms.addActions("securityhub:BatchUpdateFindings")
    sechubPerms.effect = Effect.ALLOW
    sechubPerms.addResources("*")

    memberPolicy.addStatements(sechubPerms)

    let principalPolicyStatement = new PolicyStatement();

    principalPolicyStatement.addActions("sts:AssumeRole");
    principalPolicyStatement.effect = Effect.ALLOW;

    let roleprincipal = new ArnPrincipal(
        `arn:${stack.partition}:iam::${props.adminAccountId}:role/${props.adminRoleName}`
    );

    let principals = new CompositePrincipal(roleprincipal);
    principals.addToPolicy(principalPolicyStatement);

    let serviceprincipal = new ServicePrincipal('ssm.amazonaws.com')
    principals.addPrincipals(serviceprincipal);

    let memberRole = new Role(this, 'MemberAccountRole', {
        assumedBy: principals,
        inlinePolicies: {
            'member_orchestrator': memberPolicy
        },
        roleName: `${RESOURCE_PREFIX}-SHARR-Orchestrator-Member`
    });

    const memberRoleResource = memberRole.node.findChild('Resource') as CfnRole;

    memberRoleResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W11',
                reason: 'Resource * is required due to the administrative nature of the solution.'
            },{
                id: 'W28',
                reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
            }]
        }
    }
  }
}