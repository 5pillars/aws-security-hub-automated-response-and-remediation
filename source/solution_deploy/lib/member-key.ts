// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Stack } from 'aws-cdk-lib';
import { AccountRootPrincipal, Effect, PolicyDocument, PolicyStatement, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Alias, CfnAlias, CfnKey, Key } from 'aws-cdk-lib/aws-kms';
import { CfnParameter as CfnSsmParameter, StringParameter } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';

export interface MemberKeyProps {
  readonly solutionId: string;
}

export class MemberKey extends Construct {
  constructor(scope: Construct, id: string, props: MemberKeyProps) {
    super(scope, id);

    const stack = Stack.of(this);

    const kmsKeyPolicy = new PolicyDocument();
    const kmsPerms = new PolicyStatement();
    kmsPerms.addActions(
      'kms:GenerateDataKey',
      'kms:GenerateDataKeyPair',
      'kms:GenerateDataKeyPairWithoutPlaintext',
      'kms:GenerateDataKeyWithoutPlaintext',
      'kms:Decrypt',
      'kms:Encrypt',
      'kms:ReEncryptFrom',
      'kms:ReEncryptTo',
      'kms:DescribeKey',
      'kms:DescribeCustomKeyStores'
    );
    kmsPerms.effect = Effect.ALLOW;
    kmsPerms.addResources('*'); // Only the key the policydocument is attached to
    kmsPerms.addPrincipals(new ServicePrincipal('sns.amazonaws.com'));
    kmsPerms.addPrincipals(new ServicePrincipal('s3.amazonaws.com'));
    kmsPerms.addPrincipals(new ServicePrincipal(`logs.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal(`logs.${stack.region}.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal(`cloudtrail.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal('cloudwatch.amazonaws.com'));
    kmsKeyPolicy.addStatements(kmsPerms);

    const kmsRootPolicy = new PolicyStatement({
      principals: [new AccountRootPrincipal()],
      actions: ['kms:*'],
      resources: ['*'],
    });
    kmsKeyPolicy.addStatements(kmsRootPolicy);

    const kmsKey: Key = new Key(this, 'SHARR Remediation Key', {
      enableKeyRotation: true,
      policy: kmsKeyPolicy,
    });
    (kmsKey.node.defaultChild as CfnKey).overrideLogicalId('SHARRRemediationKeyE744743D');

    const alias = new Alias(this, 'SHARR Remediation Key Alias', {
      aliasName: `${props.solutionId}-SHARR-Remediation-Key`,
      targetKey: kmsKey,
    });
    (alias.node.defaultChild as CfnAlias).overrideLogicalId('SHARRRemediationKeyAlias5531874D');

    const ssmParam = new StringParameter(this, 'SHARR Key Alias', {
      description: 'KMS Customer Managed Key that will encrypt data for remediations',
      parameterName: `/Solutions/${props.solutionId}/CMK_REMEDIATION_ARN`,
      stringValue: kmsKey.keyArn,
    });
    (ssmParam.node.defaultChild as CfnSsmParameter).overrideLogicalId('SHARRKeyAliasEBF509D8');
  }
}
