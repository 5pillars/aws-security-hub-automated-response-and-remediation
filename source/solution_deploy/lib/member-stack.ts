// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { readdirSync } from 'fs';
import { StackProps, Stack, App, CfnParameter, CfnCondition, Fn, CfnMapping, CfnStack } from 'aws-cdk-lib';
import { PolicyStatement, Effect, PolicyDocument, ServicePrincipal, AccountRootPrincipal } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Key } from 'aws-cdk-lib/aws-kms';
import { StringParameter } from 'aws-cdk-lib/aws-ssm';
import AdminAccountParam from '../../lib/admin-account-param';
import { RunbookFactory } from './runbook_factory';
import { RedshiftAuditLogging } from './redshift-audit-logging';

export interface SolutionProps extends StackProps {
  solutionId: string;
  solutionDistBucket: string;
  solutionTMN: string;
  solutionVersion: string;
  runtimePython: Runtime;
}

export class MemberStack extends Stack {
  constructor(scope: App, id: string, props: SolutionProps) {
    super(scope, id, props);

    const adminAccountParam = new AdminAccountParam(this, 'AdminAccountParameter');

    new RedshiftAuditLogging(this, 'RedshiftAuditLogging', { solutionId: props.solutionId });

    //--------------------------
    // KMS Customer Managed Key

    const stack = Stack.of(this);
    // Key Policy
    const kmsKeyPolicy: PolicyDocument = new PolicyDocument();
    const kmsPerms: PolicyStatement = new PolicyStatement();
    kmsPerms.addActions(
      'kms:GenerateDataKey',
      'kms:GenerateDataKeyPair',
      'kms:GenerateDataKeyPairWithoutPlaintext',
      'kms:GenerateDataKeyWithoutPlaintext',
      'kms:Decrypt',
      'kms:Encrypt',
      'kms:ReEncryptFrom',
      'kms:ReEncryptTo',
      'kms:DescribeKey',
      'kms:DescribeCustomKeyStores'
    );
    kmsPerms.effect = Effect.ALLOW;
    kmsPerms.addResources('*'); // Only the key the policydocument is attached to
    kmsPerms.addPrincipals(new ServicePrincipal('sns.amazonaws.com'));
    kmsPerms.addPrincipals(new ServicePrincipal('s3.amazonaws.com'));
    kmsPerms.addPrincipals(new ServicePrincipal(`logs.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal(`logs.${stack.region}.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal(`cloudtrail.${stack.urlSuffix}`));
    kmsPerms.addPrincipals(new ServicePrincipal('cloudwatch.amazonaws.com'));
    kmsKeyPolicy.addStatements(kmsPerms);

    const kmsRootPolicy = new PolicyStatement({
      principals: [new AccountRootPrincipal()],
      actions: ['kms:*'],
      resources: ['*'],
    });
    kmsKeyPolicy.addStatements(kmsRootPolicy);

    const kmsKey: Key = new Key(this, 'SHARR Remediation Key', {
      enableKeyRotation: true,
      alias: `${props.solutionId}-SHARR-Remediation-Key`,
      policy: kmsKeyPolicy,
    });

    new StringParameter(this, 'SHARR Key Alias', {
      description: 'KMS Customer Managed Key that will encrypt data for remediations',
      parameterName: `/Solutions/${props.solutionId}/CMK_REMEDIATION_ARN`,
      stringValue: kmsKey.keyArn,
    });

    new StringParameter(this, 'SHARR Member Version', {
      description: 'Version of the AWS Security Hub Automated Response and Remediation solution',
      parameterName: `/Solutions/${props.solutionId}/member-version`,
      stringValue: props.solutionVersion,
    });

    /********************
     ** Parameters
     ********************/

    const logGroupName = new CfnParameter(this, 'LogGroupName', {
      type: 'String',
      description:
        'Name of the log group to be used to create metric filters and cloudwatch alarms. You must use a Log Group that is the the logging destination of a multi-region CloudTrail',
    });

    /*********************************************
     ** Create SSM Parameter to store log group name
     *********************************************/
    new StringParameter(stack, 'SSMParameterLogGroupName', {
      description: 'Parameter to store log group name',
      parameterName: `/Solutions/${props.solutionId}/Metrics_LogGroupName`,
      stringValue: logGroupName.valueAsString,
    });

    /*********************************************
     ** Create SSM Parameter to store encryption key alias for the PCI.S3.4/AFSBP.S3.4
     *********************************************/
    new StringParameter(stack, 'SSMParameterForS3.4EncryptionKeyAlias', {
      description:
        'Parameter to store encryption key alias for the PCI.S3.4/AFSBP.S3.4, replace the default value with the KMS Key Alias, other wise the remediation will enable the default AES256 encryption for the bucket.',
      parameterName: `/Solutions/${props.solutionId}/afsbp/1.0.0/S3.4/KmsKeyAlias`,
      stringValue: 'default-s3-encryption',
    });

    new CfnMapping(this, 'SourceCode', {
      mapping: {
        General: {
          S3Bucket: props.solutionDistBucket,
          KeyPrefix: props.solutionTMN + '/' + props.solutionVersion,
        },
      },
    });

    const runbookFactory = new RunbookFactory(this, 'RunbookProvider');

    //-------------------------------------------------------------------------
    // Runbooks - shared automations
    //
    const runbookStack = new CfnStack(this, `RunbookStackNoRoles`, {
      templateUrl:
        'https://' +
        Fn.findInMap('SourceCode', 'General', 'S3Bucket') +
        '-reference.s3.amazonaws.com/' +
        Fn.findInMap('SourceCode', 'General', 'KeyPrefix') +
        '/aws-sharr-remediations.template',
    });

    runbookStack.node.addDependency(runbookFactory);

    //-------------------------------------------------------------------------
    // Loop through all of the Playbooks to create reference
    //
    const PB_DIR = `${__dirname}/../../playbooks`;
    const ignore = ['.DS_Store', 'common', '.pytest_cache', 'NEWPLAYBOOK', '.coverage'];
    const illegalChars = /[\\._]/g;
    const listOfPlaybooks: string[] = [];
    const items = readdirSync(PB_DIR);
    items.forEach((file) => {
      if (!ignore.includes(file)) {
        const templateFile = `${file}MemberStack.template`;
        //---------------------------------------------------------------------
        // Playbook Member Template Nested Stack
        //
        const parmname = file.replace(illegalChars, '');
        const memberStackOption = new CfnParameter(this, `LoadMemberStack${parmname}`, {
          type: 'String',
          description: `Load Playbook member stack for ${file}?`,
          default: 'yes',
          allowedValues: ['yes', 'no'],
        });
        memberStackOption.overrideLogicalId(`Load${parmname}MemberStack`);
        listOfPlaybooks.push(memberStackOption.logicalId);

        const memberStack = new CfnStack(this, `PlaybookMemberStack${file}`, {
          parameters: {
            SecHubAdminAccount: adminAccountParam.value,
          },
          templateUrl:
            'https://' +
            Fn.findInMap('SourceCode', 'General', 'S3Bucket') +
            '-reference.s3.amazonaws.com/' +
            Fn.findInMap('SourceCode', 'General', 'KeyPrefix') +
            '/playbooks/' +
            templateFile,
        });

        memberStack.node.addDependency(runbookFactory);

        memberStack.cfnOptions.condition = new CfnCondition(this, `load${file}Cond`, {
          expression: Fn.conditionEquals(memberStackOption, 'yes'),
        });
      }
    });

    /********************
     ** Metadata
     ********************/
    stack.templateOptions.metadata = {
      'AWS::CloudFormation::Interface': {
        ParameterGroups: [
          {
            Label: { default: 'LogGroup Configuration' },
            Parameters: [logGroupName.logicalId],
          },
          {
            Label: { default: 'Playbooks' },
            Parameters: listOfPlaybooks,
          },
        ],
        ParameterLabels: {
          [logGroupName.logicalId]: {
            default: 'Provide the name of the LogGroup to be used to create Metric Filters and Alarms',
          },
        },
      },
    };
  }
}
