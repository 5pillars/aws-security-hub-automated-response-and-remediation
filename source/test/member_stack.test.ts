// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { App, Aspects, DefaultStackSynthesizer, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AwsSolutionsChecks } from 'cdk-nag';
import { MemberStack } from '../solution_deploy/lib/member-stack';

function getCatStack(): Stack {
  const app = new App();
  const stack = new MemberStack(app, 'CatalogStack', {
    synthesizer: new DefaultStackSynthesizer({ generateBootstrapVersionRule: false }),
    description: 'test;',
    solutionId: 'SO0111',
    solutionVersion: 'v1.1.1',
    solutionDistBucket: 'sharrbukkit',
    solutionTMN: 'aws-security-hub-automated-response-and-remediation',
    runtimePython: Runtime.PYTHON_3_9,
  });
  Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));
  return stack;
}

test('default stack', () => {
  expect(Template.fromStack(getCatStack())).toMatchSnapshot();
});
