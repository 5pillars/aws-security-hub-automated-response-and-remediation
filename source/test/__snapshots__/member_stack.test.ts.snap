// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default stack 1`] = `
{
  "Conditions": {
    "EnableS3BucketForRedShift4": {
      "Fn::Equals": [
        {
          "Ref": "CreateS3BucketForRedshiftAuditLogging",
        },
        "yes",
      ],
    },
  },
  "Description": "test;",
  "Mappings": {
    "SourceCode": {
      "General": {
        "KeyPrefix": "aws-security-hub-automated-response-and-remediation/v1.1.1",
        "S3Bucket": "sharrbukkit",
      },
    },
  },
  "Metadata": {
    "AWS::CloudFormation::Interface": {
      "ParameterGroups": [
        {
          "Label": {
            "default": "LogGroup Configuration",
          },
          "Parameters": [
            "LogGroupName",
          ],
        },
        {
          "Label": {
            "default": "Playbooks",
          },
          "Parameters": [],
        },
      ],
      "ParameterLabels": {
        "LogGroupName": {
          "default": "Provide the name of the LogGroup to be used to create Metric Filters and Alarms",
        },
      },
    },
  },
  "Parameters": {
    "CreateS3BucketForRedshiftAuditLogging": {
      "AllowedValues": [
        "yes",
        "no",
      ],
      "Default": "no",
      "Description": "Create S3 Bucket For Redshift Cluster Audit Logging.",
      "Type": "String",
    },
    "LogGroupName": {
      "Description": "Name of the log group to be used to create metric filters and cloudwatch alarms. You must use a Log Group that is the the logging destination of a multi-region CloudTrail",
      "Type": "String",
    },
    "SecHubAdminAccount": {
      "AllowedPattern": "\\d{12}",
      "Description": "Admin account number",
      "Type": "String",
    },
  },
  "Resources": {
    "RunbookStackNoRoles": {
      "Properties": {
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://",
              {
                "Fn::FindInMap": [
                  "SourceCode",
                  "General",
                  "S3Bucket",
                ],
              },
              "-reference.s3.amazonaws.com/",
              {
                "Fn::FindInMap": [
                  "SourceCode",
                  "General",
                  "KeyPrefix",
                ],
              },
              "/aws-sharr-remediations.template",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
    },
    "S3BucketForRedShiftAuditLogging652E7355": {
      "Condition": "EnableS3BucketForRedShift4",
      "DeletionPolicy": "Retain",
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-S1",
              "reason": "Logs bucket does not require logging configuration",
            },
          ],
        },
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W35",
              "reason": "Logs bucket does not require logging configuration",
            },
          ],
        },
      },
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "S3BucketForRedShiftAuditLoggingBucketPolicyAB8BAA40": {
      "Condition": "EnableS3BucketForRedShift4",
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "S3BucketForRedShiftAuditLogging652E7355",
      ],
      "Properties": {
        "Bucket": {
          "Ref": "S3BucketForRedShiftAuditLogging652E7355",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetBucketAcl",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "redshift.amazonaws.com",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "S3BucketForRedShiftAuditLogging652E7355",
                    "Arn",
                  ],
                },
                {
                  "Fn::Sub": [
                    "arn:\${AWS::Partition}:s3:::\${BucketName}/*",
                    {
                      "BucketName": {
                        "Ref": "S3BucketForRedShiftAuditLogging652E7355",
                      },
                    },
                  ],
                },
              ],
              "Sid": "Put bucket policy needed for audit logging",
            },
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": "*",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "S3BucketForRedShiftAuditLogging652E7355",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "S3BucketForRedShiftAuditLogging652E7355",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
              "Sid": "EnforceSSL",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
      "UpdateReplacePolicy": "Retain",
    },
    "SHARRKeyAliasEBF509D8": {
      "Properties": {
        "Description": "KMS Customer Managed Key that will encrypt data for remediations",
        "Name": "/Solutions/SO0111/CMK_REMEDIATION_ARN",
        "Type": "String",
        "Value": {
          "Fn::GetAtt": [
            "SHARRRemediationKeyE744743D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SHARRMemberVersionEDAB5C42": {
      "Properties": {
        "Description": "Version of the AWS Security Hub Automated Response and Remediation solution",
        "Name": "/Solutions/SO0111/member-version",
        "Type": "String",
        "Value": "v1.1.1",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SHARRRemediationKeyAlias5531874D": {
      "Properties": {
        "AliasName": "alias/SO0111-SHARR-Remediation-Key",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "SHARRRemediationKeyE744743D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "SHARRRemediationKeyE744743D": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "EnableKeyRotation": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": [
                "kms:GenerateDataKey",
                "kms:GenerateDataKeyPair",
                "kms:GenerateDataKeyPairWithoutPlaintext",
                "kms:GenerateDataKeyWithoutPlaintext",
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncryptFrom",
                "kms:ReEncryptTo",
                "kms:DescribeKey",
                "kms:DescribeCustomKeyStores",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": [
                  "sns.amazonaws.com",
                  "s3.amazonaws.com",
                  {
                    "Fn::Join": [
                      "",
                      [
                        "logs.",
                        {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                  {
                    "Fn::Join": [
                      "",
                      [
                        "logs.",
                        {
                          "Ref": "AWS::Region",
                        },
                        ".",
                        {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                  {
                    "Fn::Join": [
                      "",
                      [
                        "cloudtrail.",
                        {
                          "Ref": "AWS::URLSuffix",
                        },
                      ],
                    ],
                  },
                  "cloudwatch.amazonaws.com",
                ],
              },
              "Resource": "*",
            },
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "SSMParameterForS34EncryptionKeyAlias73DD8A98": {
      "Properties": {
        "Description": "Parameter to store encryption key alias for the PCI.S3.4/AFSBP.S3.4, replace the default value with the KMS Key Alias, other wise the remediation will enable the default AES256 encryption for the bucket.",
        "Name": "/Solutions/SO0111/afsbp/1.0.0/S3.4/KmsKeyAlias",
        "Type": "String",
        "Value": "default-s3-encryption",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SSMParameterForS3BucketNameForREDSHIFT441DD36B1": {
      "Condition": "EnableS3BucketForRedShift4",
      "DependsOn": [
        "S3BucketForRedShiftAuditLogging652E7355",
      ],
      "Properties": {
        "Description": "Parameter to store the S3 bucket name for the remediation AFSBP.REDSHIFT.4, the default value is bucket-name which has to be updated by the user before using the remediation.",
        "Name": "/Solutions/SO0111/afsbp/1.0.0/REDSHIFT.4/S3BucketNameForAuditLogging",
        "Type": "String",
        "Value": {
          "Ref": "S3BucketForRedShiftAuditLogging652E7355",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SSMParameterLogGroupName47918519": {
      "Properties": {
        "Description": "Parameter to store log group name",
        "Name": "/Solutions/SO0111/Metrics_LogGroupName",
        "Type": "String",
        "Value": {
          "Ref": "LogGroupName",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;
